### ISSUE TITLE
How to have multiple inputs to a DAG network?

### ISSUE THREAD

**Issue by luzhi** [vlfeat/matconvnet#553]
I have the images for one object with multiple views (e.g., 2 views for one object). In the proposed DAG network, the network structure looks like this:

1. Two linear chain (simple CNN) networks accept the data of the 2 views respectively.
2. There is a fusion layer (FC-layer) to concatenate the features of these two data sources by FC layers in each network.
3. A loss function is followed by the previous fusion layer.

My question is how to have the inputs with multiple views of one object to the two simple CNNs in the whole DAG network? I tried to learn from the example code "cnn_train_dag.m", but still have no idea about this. Could you please advise me on that? Thanks!

**Comment by luzhi** [comment]
There are two points we need to consider when we build up a DAG network that accepts 2 or more inputs:

**1. inputs = getDagNNBatch(opts, imdb, batch)**
This function is defined to build up the input data, which has one or more types of data, to the DAG from the _imdb_. It returns a cell-array, called **inputs**, in which it is organized like this: 
`inputs = {'string_name_of_input_1', inputs_variable_1, 'labels_1', labels_1, ...`
               `'string_name_of_input_2, inputs_variable_2, 'labels_2', labels_2};`

Depending on the number of the input data we have, we can adapt this function _getDagNNBatch(...)_ to our application:

```
% -------------------------------------------------------------------------
function inputs = getDagNNBatch(opts, imdb, batch)
% -------------------------------------------------------------------------
images_src_1 = imdb.images.data_src1(:,:,:,batch) ;
images_src_2 = imdb.images.data_src2(:,:,:,batch) ;
labels = imdb.images.labels(1,batch) ;
if rand > 0.5
    images_src_1 =fliplr(images_src_1 ) ; 
    images_src_2 =fliplr(images_src_2 ) ; 
end
if opts.numGpus > 0
  images_src_1 = gpuArray(images_src_1 ) ;
  images_src_2 = gpuArray(images_src_2 ) ;
end
% **********************************************
% Define the inputs cell-array to the DAG
% **********************************************
inputs = {'input_1', images_src_1 , 'label', labels, 'input_2', images_src_2, 'label', labels} ;
```

This example gives how to have two types of input data to the same DAG network and there is only one labels to the network. This function will be called when you train the network to get the bathes of data from both data sources.

It is necessary to generate different types of input data when you build up the _imdb_ variable.

**2. Build up the first layer of the DAG network**
There is a code snippet on the definition of the input layers in the DAG network. For example, we have two types of input data as above:
```
% init the object:
nn = dagnn.DagNN();
  
% the DAG network structure  
% ------------------------------------------------------------------
% BLOCK 1
% ------------------------------------------------------------------  
% First layer accepts the first input data
l_cnv1 = dagnn.Conv('size', [5 5 3 32], 'pad', 2, 'stride', 1, 'hasBias', true);
nn.addLayer('l_cnv1 ', l_cnv1 , {'input_1'}, {'x1_1'}, {'l1f', 'l1b'});

l_cnv2 = dagnn.Conv('size', [3 3 3 64], 'pad', 2, 'stride', 1, 'hasBias', true);
nn.addLayer('l_cnv2 ', l_cnv2 , {'input_2'}, {'x1_2'}, {'l2f', 'l2b'});
```

**Comment by yushuinanrong** [comment]
@luzhi 
Hi luzhi,

I'm also trying to implement two linear CNNs sharing fully connected layers in Matcovnet. I'm wondering if you could share the complete script doing that?

regards,
jiawei

**Comment by h612** [comment]
But how to concatinate two networks? One network is Net1 the other with different input is Net2. So if Net is the final output after concatination.. it doesn't include either of the two networks..and only displays FC layers.. :<
